bin/cml/comment/create.js

```diff
@@ -0,0 +1,53 @@
+const kebabcaseKeys = require('kebabcase-keys');
+
+exports.command = 'create <markdown file>';
+exports.description = 'Create a comment';
+
+exports.handler = async (opts) => {
+  const { cml } = opts;
+  console.log(await cml.commentCreate(opts));
+};
+
+exports.builder = (yargs) => yargs.env('CML_COMMENT').options(exports.options);
+
+exports.options = kebabcaseKeys({
+  pr: {
+    type: 'boolean',
+    description:
+      'Post to an existing PR/MR associated with the specified commit'
+  },
+  commitSha: {
+    type: 'string',
+    alias: 'head-sha',
+    default: 'HEAD',
+    description: 'Commit SHA linked to this comment'
+  },
+  publish: {
+    type: 'boolean',
+    description: 'Upload local images which are inlined in the Markdown comment'
+  },
+  watch: {
+    type: 'boolean',
+    description: 'Watch for changes and automatically update the comment'
+  },
+  triggerFile: {
+    type: 'string',
+    description: 'File used to trigger the watcher',
+    hidden: true
+  },
+  native: {
+    type: 'boolean',
+    description:
+      "Uses driver's native capabilities to upload assets instead of CML's storage; not available on GitHub"
+  },
+  update: {
```

Comment:
Should there be a `cml comment update` alias `cml comment create --update`?

Reply:
See https://github.com/iterative/cml/issues/762#issuecomment-1163503318 and subsequent comments for more insight on this choice. Won't be part of this pull request, but might make sense.

Reply:
maybe we should make a follow-up issue?

Reply:
Feel free to 👍🏼

Reply:
> See [#762 (comment)](https://github.com/iterative/cml/issues/762#issuecomment-1163503318) and subsequent comments

- deal-breaker for me as per every single "subsequent comment." Definitely not for a follow-up, must be addressed here
  + the primary objective of [#1073](https://github.com/iterative/cml/pull/1073) (<- [#762](https://github.com/iterative/cml/issues/762), [#1026](https://github.com/iterative/cml/pull/1026)) is fixing this particular subcommand's UX. Everything else is comparatively lower-priority technical debt.
- as per https://github.com/iterative/cml/pull/1026#discussion_r901479836, definitely must be one of:
  1. `cml report [--pr,--publish [--native]]`[^native]
     + `create [--rm-watermark]`
     + `update`
     + `watch [{--rm-watermark,--update},--trigger-file]`[^watermark]
  2. `cml report [--pr,--publish [--native],--watch [--trigger-file]]`[^native][^trigger]
     + `create [--rm-watermark]`
     + `update`

Meanwhile, `cml comment create --update` is bad design because it breaks [`CRUD`](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) (`U` is meant to be at the same level as `C`, not a subset). It implies a future with the hilarious `cml comment create --delete` (on par with "[click `start` to shutdown](https://devblogs.microsoft.com/oldnewthing/20030722-00/?p=43083)").

[^watermark]: `--rm-watermark` conflicts with `--update`
[^trigger]: `--trigger-file` requires `--watch`
[^native]: `--native` requires `--publish`

Reply:
The `cml <noun> create --update` invocation has a very specific meaning:

1. `cml <noun> create` — create a new `<noun>` or fail if one already exists
2. `cml <noun> update` — update an existing `<noun>` or fail if none was found
3. `cml <noun> create --update` — create a new `<noun>` if none was found, else update an existing one

> **Note**
> To complicate things a bit more, updates are always performed over the CML comment with the latest timestamp, and users have no control over the comment they want to update.

Currently, `cml` always behaves in mode **3**. Official tools like `gh` provide just modes **1** and **2**, relying on the users' judgement to determine whet trusting users whether to create or update and what to update.

Reply:
With regard to breaking `CRUD` and actions being at the same level, please note that it's a widespread practice to provide shorthands in command-line tools to merge several verbs into a single invocation.

### Example

#### Verbs
* `command <noun> <verb-1>`
* `command <noun> <verb-2>`

#### Shorthands
* `command <noun> <verb-1> [--verb-2]`
* `command <noun> <verb-2> [--verb-1]`

### References

* [`git checkout -b` — _causes a new branch to be created as if `git-branch` were called and then checked out_](https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt-emgitcheckoutem-b-Bltnew-branchgtltstart-pointgt)

Reply:
Nitpick: the `--rm-watermark` option should be `--no-watermark` if anything, speaking about user experience.

Reply:
> _the primary objective of https://github.com/iterative/cml/pull/1073 (<- https://github.com/iterative/cml/issues/762, https://github.com/iterative/cml/pull/1026) is fixing this particular subcommand's UX_

As per the body of #1073 (this pull request), _“[it] doesn't quite close #762”_

Reply:
> _must be addressed here_

Then, I'll start watching beard shaving tutorials.

Reply:
sure, `rm-watermark` could be `no-watermark`, and `update` could be `try-update-fallback-create`. I don't really mind what they're called, open to suggestions. I do mind whether they are `commands` or `--options` though. It's about use cases:

current | proposals (i) or (ii) | use case
---|---|---
`cml comment create --update`| `cml comment update` | one experiment per PR (similar use case to `check`s but uncollapsed report. Similar also to code coverage reports)
`cml comment create` | `cml comment create` | one experiment per commit (multiple experiments per PR)

The `update` use case is far too important to be relegated to an `--option`.

Reply:
I can see the use of the --update flag. ex ci triggered on `pull_request` gets re-run every time the user pushes more commits.

I wouldn't want to the have logic that tries to determine if its the first run to use `create` else `update` additionally I wouldn't want to create multiple comments of essentially the same data with small changes.

I think that `cml comment update` should exist but I agree that something analogous to the `git checkout -b` metaphor should be present.
| example | action |
| --- | --- |
| `cml comment create` | "normal" always creates a comment |
| `cml comment create -u` | looks for watermark -> update else create |
| `cml comment update` | looks for watermark -> updates |
| `cml comment update --id 1234` | tries to update specific comment |

perhaps stealing some HTTP terminology or is that too ugly? `cml comment put`

Reply:
Hmm, taking a step back (ignoring how the choices are specified, i.e. via `subcommand` or via `--option`), this is my reasoning: there are some choices which are nonsensical, as per the table below. Do we agree?

create | update | rm-watermark | watch | behaviour
-|-|-|-|-
1|0|0|0|new
0|1|0|0|mod existing (:warning: personally would prefer "fallback create" as I see zero use case for throwing an error)
1|1|0|0|mod existing, fallback create
1|0|1|0|new, read-only (non-updatable)
0|1|1|0|error
1|1|1|0|mod existing & convert to read-only, fallback create read-only
1|0|0|1|multiple new (:warning: personally would prefer error)
0|1|0|1|mod existing (:warning: personally would prefer "fallback create" as I see zero use case for throwing an error), multiple subsequent mods
1|1|0|1|mod existing, fallback create, multiple subsequent mods
1|0|1|1|error
0|1|1|1|error
1|1|1|1|error

Reply:
Indeed, some of them are either nonsensical or useless for the workflows we recommend.

To my mind, the following constraints would simplify considerably the workflow:
1. `watch` always implies `update`
2. `update` always implies `create`
3. `--rm-watermark` shouldn't be part of our public API, unless we have a clear use case

Reply:
This conversation is probably related to https://github.com/iterative/cml/pull/1073#discussion_r957923322. Currently all of our commands have a single verb, loosely equivalent to the CU[^1] in CRUD.

Reply:
> `watch` always implies `update` [and] `update` always implies `create`

individually, those are fine assumptions. However they can't be combined: there's 2 subtly different use cases for `watch`: a) update existing (no create) and b) create new and keep updating it... meaning `watch` does not always imply `create`.

As I see it, there are 4 use cases, one of which isn't even implemented:

use case | current | proposed (?) | proposed (casper)
-|-|-|-
(1) create comment | `create` | `create` | `create`
(2) update existing, fallback create | `create --update` | `create --update` | `update`
(3) create comment & keep modding it | :warning: not implemented | `create --watch` | `create --watch`
(4) keep modding existing | `create --watch` | `create --watch --update` | `update --watch`
(?) update existing, error if not found | not implemented | `update` | not implemented[^grudge]

Makes sense now? Also [`create --update` seems as illogical to me as `s/workflow rerun/workflow run --rerun/`](https://github.com/iterative/cml/issues/762#issuecomment-1163570557)

> `--rm-watermark` shouldn't be part of our public API

sounds like a follow-up issue; added to OP.

[^grudge]: grudgingly `update --no-create` if users request this functionality

Reply:
Isn't 3 implemented? 🤔

Reply:
With regard to (2), isn't it `update --create-if-does-not-exist`? 🤔 Also known as your favorite `update --create` or `create --update` 😅

Reply:
Sounds more reasonable to me than `create --delete` or `delete --update` 😸

Reply:
Resolved with https://github.com/iterative/cml/pull/1073/commits/aa48409ff0d1e02a3951773cda29aaec54f49362